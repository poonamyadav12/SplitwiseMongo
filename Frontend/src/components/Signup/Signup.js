import React, { Component } from 'react';
import { Button, Card, Col, Container, Form, ListGroup, Row } from 'react-bootstrap';
import { Typeahead } from 'react-bootstrap-typeahead';
import { connect } from 'react-redux';
import { Redirect } from 'react-router';
import '../../App.css';
import { userActions } from '../../_actions';
import { getDefaultUserImage } from '../../_constants/avatar';
import { CURRENCY } from '../../_helper/money';
import { UploadImage } from '../Image/UploadImage';
import TimezoneSelect from 'react-timezone-select';

//Define a Login Component
class Signup extends Component {
    //call the constructor method
    constructor(props) {
        //Call the constrictor of Super class i.e The Component
        super(props);
        //maintain the state required for this component
        this.state = {
            firstName: "",
            lastName: "",
            username: "",
            password: "",
            errorMsg: "",
            timeZone: null,
            currency: 'USD',
            imageUrl: "",
        }

        //Bind the handlers to this class
        this.firstNameChangeHandler = this.firstNameChangeHandler.bind(this);
        this.lastNameChangeHandler = this.lastNameChangeHandler.bind(this);
        this.usernameChangeHandler = this.usernameChangeHandler.bind(this);
        this.passwordChangeHandler = this.passwordChangeHandler.bind(this);
        this.submitSignup = this.submitSignup.bind(this);
        this.timezoneHandler = this.timezoneHandler.bind(this);
        this.currencyHandler = this.currencyHandler.bind(this);
        this.avatarChangeHandler = this.avatarChangeHandler.bind(this);
    }
    //Call the Will Mount to set the auth Flag to false
    componentWillMount() {
        this.setState({
            authFlag: false
        })
    }
    avatarChangeHandler = (imageUrl) => {
        this.setState({ imageUrl });
    }

    //name change handler to update state variable with the text entered by the user
    firstNameChangeHandler = (e) => {
        this.setState({
            firstName: e.target.value
        })
    }
    //name change handler to update state variable with the text entered by the user
    lastNameChangeHandler = (e) => {
        this.setState({
            lastName: e.target.value
        })
    }
    //username change handler to update state variable with the text entered by the user
    usernameChangeHandler = (e) => {
        this.setState({
            username: e.target.value
        })
    }
    //password change handler to update state variable with the text entered by the user
    passwordChangeHandler = (e) => {
        this.setState({
            password: e.target.value
        })
    }
    //timezone change handler
    timezoneHandler = (value) => {
        this.setState({
            timeZone: value.value
        })
    }
    //timezone change handler
    currencyHandler = (value) => {
        this.setState({
            currency: value[0].name
        })
    }
    setHasError = (errorMsg) => {
        this.setState({
            errorMsg: errorMsg
        })
    }
    //submit Login handler to send a request to the node backend
    submitSignup = (e) => {
        //var headers = new Headers();
        //prevent page from refresh
        e.preventDefault();
        const data = {
            user: {
                first_name: this.state.firstName,
                last_name: this.state.lastName,
                email: this.state.username,
                default_currency: this.state.currency,
                password: this.state.password,
                time_zone: this.state.timeZone,
                avatar: this.state.imageUrl,

            }
        }

        this.props.register(data);
    }
    buildErrorComponent = () => {
        return this.state.errorMsg && <div class="alert alert-danger" role="alert">{this.state.errorMsg}</div>
    }
    render() {
        //redirect based on successful login
        return (
            <>
                {this.props.user && <Redirect to="/home" />}
                <Container>
                    <Row style={{ alignItems: 'center' }}>
                        <Col lg={3} style={{ 'width': '20%', marginTop: '3%' }}>
                            <img style={{ width: '17rem' }} src="https://assets.splitwise.com/assets/core/logo-square-65a6124237868b1d2ce2f5db2ab0b7c777e2348b797626816400534116ae22d7.svg" alt="Logo" />
                        </Col>
                        <Col lg={3}>
                            <UploadImage defaultImage={getDefaultUserImage()} onChange={this.avatarChangeHandler}></UploadImage>
                        </Col>
                        <Col lg={6}>
                            <Card style={{ width: '35rem' }}>
                                <Form>
                                    {this.buildErrorComponent()}
                                    <Form.Group controlId="formBasicName">
                                        <Form.Label>First name</Form.Label>
                                        <Form.Control type="text" placeholder="Enter first name" onChange={this.firstNameChangeHandler} />
                                    </Form.Group>
                                    <Form.Group controlId="formBasicLastName">
                                        <Form.Label>Last name</Form.Label>
                                        <Form.Control type="text" placeholder="Enter last name" onChange={this.lastNameChangeHandler} />
                                    </Form.Group>
                                    <Form.Group controlId="formBasicEmail">
                                        <Form.Label>Email address</Form.Label>
                                        <Form.Control type="email" placeholder="Enter email" onChange={this.usernameChangeHandler} />
                                    </Form.Group>
                                    <Form.Group controlId="formBasicPassword">
                                        <Form.Label>Password</Form.Label>
                                        <Form.Control type="password" placeholder="Password" onChange={this.passwordChangeHandler} />
                                    </Form.Group>
                                    <Form.Group>
                                        <Form.Label>Currency</Form.Label>
                                        <Typeahead
                                            id="currency"
                                            labelKey="name"
                                            onChange={this.currencyHandler}
                                            options={CURRENCY}
                                            placeholder="Choose a currency..."
                                            renderMenuItemChildren={(option, props) => (
                                                <ListGroup.Item key={option.name}>
                                                    <span>{option.name}</span>{' '}
                                                </ListGroup.Item>
                                            )}
                                        />
                                    </Form.Group>
                                    <Form.Group>
                                        <Form.Label>Time Zone</Form.Label>
                                        <TimezoneSelect
                                            value={'America/Los Angeles'}
                                            onChange={this.timezoneHandler}
                                            timezones={{
                                                'America/Los_Angeles': 'Los Angeles',
                                                'Europe/Berlin': 'Frankfurt',
                                                'Asia/Kuwait': 'Kuwait',
                                                'Asia/Bahrain': 'Bahrain',
                                                'Canada/Mountain': 'Canada Mountain',
                                                'Canada/Pacific': 'Canada Pacific',
                                                'Asia/Kolkata': 'Indian Standard Time',
                                                //    ...i18nTimezones,
                                            }}
                                        />
                                    </Form.Group>
                                    <Button variant="success" type="submit" onClick={this.submitSignup}>
                                        Submit
                                </Button>
                                </Form>
                            </Card>
                        </Col>
                    </Row>
                </Container>
            </>
        )
    }
}

function mapState(state) {
    const { user } = state.authentication;
    const { registering } = state.registration;
    return { registering, user };
}

const actionCreators = {
    register: userActions.register
}

const connectedRegisterPage = connect(mapState, actionCreators)(Signup);
export { connectedRegisterPage as Signup, Signup as SignupForTest };
//export Login Component