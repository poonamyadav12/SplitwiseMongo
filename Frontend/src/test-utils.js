// test-utils.js
import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'
// Import your own reducer
import rootReducer from '../src/reducers';
import { createLogger } from 'redux-logger';
import { createStore, applyMiddleware, compose } from 'redux';
import thunkMiddleware from 'redux-thunk';

jest.mock('react-router-dom'); // <- Use this line
jest.mock('react-avatar-editor');
jest.mock('axios');

const loggerMiddleware = createLogger();
const composeEnhancers = (typeof window !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

function render(
  ui,
  {
    initialState,
    store = createStore(
      rootReducer,
      composeEnhancers(
        applyMiddleware(
          thunkMiddleware,
          loggerMiddleware
        ))
    ),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }