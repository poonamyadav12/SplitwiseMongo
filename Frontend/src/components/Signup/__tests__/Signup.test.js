import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import mockAxios from "axios";
import { setupServer } from 'msw/node';
import React from 'react';
import { fireEvent, render, screen } from '../../../test-utils';
import { Signup, SignupForTest } from '../Signup';
import renderer from 'react-test-renderer';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';


const server = setupServer();

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

test('displays signup page and makes API call', async () => {
  const user = {
    "user": {
      "email": "abc@gmail.com",
      "first_name": "Michael",
      "last_name": "Clarke",
      "password": "Abc@1234",
    }
  };
  mockAxios.post.mockImplementationOnce(() =>
    Promise.resolve({
      data: user
    })
  );

  render(<Signup />)

  const firstName = screen.getByPlaceholderText('Enter first name');
  fireEvent.change(firstName, { target: { value: 'Michael' } });

  const lastName = screen.getByPlaceholderText('Enter last name');
  fireEvent.change(firstName, { target: { value: 'Clarke' } });

  const email = screen.getByPlaceholderText('Enter email');
  fireEvent.change(email, { target: { value: 'abc@gmail.com' } });

  const password = screen.getByPlaceholderText('Password');
  fireEvent.change(password, { target: { value: 'Abc@1234' } });

  fireEvent.click(screen.getByText('Submit'));

  setTimeout(() => {
    expect(mockAxios.post).toHaveBeenCalledWith("http://localhost:3001/user/signup", {
      user: {
        "email": "abc@gmail.com",
        "first_name": "Michael",
        "last_name": "Clarke",
        "password": "Abc@1234",
      }
    });
  }, 100);
});

test('renders correctly', () => {
  const tree = renderer
    .create(<SignupForTest user={null} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

