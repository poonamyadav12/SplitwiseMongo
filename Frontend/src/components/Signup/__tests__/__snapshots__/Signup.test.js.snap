// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <div
      className="col-lg-3"
      style={
        Object {
          "marginTop": "3%",
          "width": "20%",
        }
      }
    >
      <img
        alt="Logo"
        src="https://assets.splitwise.com/assets/core/logo-square-65a6124237868b1d2ce2f5db2ab0b7c777e2348b797626816400534116ae22d7.svg"
        style={
          Object {
            "width": "17rem",
          }
        }
      />
    </div>
    <div
      className="col-lg-3"
    >
      <div
        className="col"
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="imageUplaod"
          >
            <h5>
              Avatar
            </h5>
          </label>
          <div
            onBlur={[Function]}
            onClick={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
          >
            <input
              autoComplete="off"
              multiple={true}
              onChange={[Function]}
              onClick={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              tabIndex={-1}
              type="file"
            />
          </div>
          <h5>
            Drop images or click to upload
          </h5>
        </div>
      </div>
    </div>
    <div
      className="col-lg-6"
    >
      <div
        className="card"
        style={
          Object {
            "width": "35rem",
          }
        }
      >
        <form
          className=""
        >
          
          <div
            className="form-group"
          >
            <label
              className="form-label"
              htmlFor="formBasicName"
            >
              First name
            </label>
            <input
              className="form-control"
              id="formBasicName"
              onChange={[Function]}
              placeholder="Enter first name"
              type="text"
            />
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
              htmlFor="formBasicLastName"
            >
              Last name
            </label>
            <input
              className="form-control"
              id="formBasicLastName"
              onChange={[Function]}
              placeholder="Enter last name"
              type="text"
            />
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
              htmlFor="formBasicEmail"
            >
              Email address
            </label>
            <input
              className="form-control"
              id="formBasicEmail"
              onChange={[Function]}
              placeholder="Enter email"
              type="email"
            />
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
              htmlFor="formBasicPassword"
            >
              Password
            </label>
            <input
              className="form-control"
              id="formBasicPassword"
              onChange={[Function]}
              placeholder="Password"
              type="password"
            />
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Currency
            </label>
            <div
              className="rbt"
              style={
                Object {
                  "outline": "none",
                  "position": "relative",
                }
              }
              tabIndex={-1}
            >
              <div
                style={
                  Object {
                    "display": "flex",
                    "flex": 1,
                    "height": "100%",
                    "position": "relative",
                  }
                }
              >
                <input
                  aria-autocomplete="both"
                  aria-expanded={false}
                  aria-haspopup="listbox"
                  autoComplete="off"
                  className="rbt-input-main form-control rbt-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Choose a currency..."
                  role="combobox"
                  type="text"
                  value=""
                />
                <input
                  aria-hidden={true}
                  className="rbt-input-hint"
                  readOnly={true}
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "borderColor": "transparent",
                      "boxShadow": "none",
                      "color": "rgba(0, 0, 0, 0.35)",
                      "left": 0,
                      "pointerEvents": "none",
                      "position": "absolute",
                      "top": 0,
                      "width": "100%",
                    }
                  }
                  tabIndex={-1}
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Time Zone
            </label>
            <div
              className=" css-2b097c-container"
              onKeyDown={[Function]}
            >
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                className="css-1f43avz-a11yText-A11yText"
              />
              <div
                className=" css-yk16xz-control"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <div
                  className=" css-g1d714-ValueContainer"
                >
                  <div
                    className=" css-1wa3eu0-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    className="css-b8ldur-Input"
                  >
                    <div
                      className=""
                      style={
                        Object {
                          "display": "inline-block",
                        }
                      }
                    >
                      <input
                        aria-autocomplete="list"
                        autoCapitalize="none"
                        autoComplete="off"
                        autoCorrect="off"
                        disabled={false}
                        id="react-select-3-input"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        spellCheck="false"
                        style={
                          Object {
                            "background": 0,
                            "border": 0,
                            "boxSizing": "content-box",
                            "color": "inherit",
                            "fontSize": "inherit",
                            "label": "input",
                            "opacity": 1,
                            "outline": 0,
                            "padding": 0,
                            "width": "1px",
                          }
                        }
                        tabIndex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style={
                          Object {
                            "height": 0,
                            "left": 0,
                            "overflow": "scroll",
                            "position": "absolute",
                            "top": 0,
                            "visibility": "hidden",
                            "whiteSpace": "pre",
                          }
                        }
                      >
                        
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    className=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    className=" css-tlfecz-indicatorContainer"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-tj5bde-Svg"
                      focusable="false"
                      height={20}
                      viewBox="0 0 20 20"
                      width={20}
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <button
            className="btn btn-success"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
